sources/file-system/src/main/java/io/smallrye/config/source/file/FileSystemConfigSourceFactory.java

```diff
@@ -26,8 +29,26 @@
 
         return Stream
                 .of(newArrayConverter(Converters.getImplicitConverter(String.class), String[].class)
-                        .convert(value.getValue()))
-                .map(location -> new FileSystemConfigSource(location))
+                            .convert(value.getValue()))
+                .flatMap(location -> {
+                             if (location.endsWith(WILDCARD_LOCATION_SUFFIX)) {
+                                 final URI rootUrl = URI.create(location.substring(0, location.length() - WILDCARD_LOCATION_SUFFIX.length()));
+                                 final String schema = rootUrl.getScheme();
+                                 if (schema != null && !"file".equalsIgnoreCase(schema)) {
+                                     throw new IllegalArgumentException("Wildcard source file location not supported for URL schema " + schema);
+                                 }
+                                 final File rootLocation = new File(rootUrl.getPath());
+                                 if (rootLocation.isDirectory()) {
+                                     return Stream.of(rootLocation.listFiles(pathname -> pathname.isDirectory()))
+                                                  .map(subLocation -> new FileSystemConfigSource(subLocation));
+                                 } else {
+                                     return Stream.empty();
+                                 }
+                             } else {
+                                 return Stream.of(new FileSystemConfigSource(location));
+                             }
+                         }
+                )
```

Comment:
I believe the logic should move to `FileSystemConfigSource` because it can be used on its own. If we keep this logic here, then you have different behaviors between `FileSystemConfigSource` and `FileSystemConfigSourceFactory`. On the other hand, `FileSystemConfigSourceFactory` will use the sources produced by `FileSystemConfigSource`, so the behaviors will be aligned.

Also, I would recommend implementing the logic with https://docs.oracle.com/javase/8/docs/api/index.html?java/nio/file/PathMatcher.html.

Reply:
> We should support `URI` for sure. You can probably check

We should fix that in this PR or file a separate PR for URI support?

Reply:
You can do everything here. Thanks!